firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/internal/persistence/CrashlyticsReportPersistence.java

```diff
@@ -0,0 +1,236 @@
+        writeTextFile(
+            new File(outputDirectory, sessionId),
+            TRANSFORM.reportToJson(report.withEvents(ImmutableList.from(events))));
+      }
+      recursiveDelete(sessionDirectory);
+    }
+  }
+
+  public List<CrashlyticsReport> loadFinalizedReports() {
+    final List<CrashlyticsReport> reports = new ArrayList<>();
+    final List<File> fatalReports = getAllFilesInDirectory(fatalReportsDirectory);
+    for (File reportFile : fatalReports) {
+      reports.add(TRANSFORM.reportFromJson(readTextFile(reportFile)));
+    }
+    final List<File> nonFatalReports = getAllFilesInDirectory(nonFatalReportsDirectory);
+    for (File reportFile : nonFatalReports) {
+      reports.add(TRANSFORM.reportFromJson(readTextFile(reportFile)));
+    }
+    return reports;
+  }
+
+  private File getSessionDirectoryById(String sessionId) {
+    return new File(openSessionsDirectory, sessionId);
+  }
+
+  private static List<File> getAllFilesInDirectory(File directory) {
+    return getFilesInDirectory(directory, (FileFilter) null);
+  }
+
+  private static List<File> getFilesInDirectory(File directory, FilenameFilter filter) {
+    if (directory == null || !directory.isDirectory()) {
+      return Collections.emptyList();
+    }
+    final File[] files = (filter == null) ? directory.listFiles() : directory.listFiles(filter);
+    return (files != null) ? Arrays.asList(files) : Collections.emptyList();
+  }
+
+  private static List<File> getFilesInDirectory(File directory, FileFilter filter) {
+    if (directory == null || !directory.isDirectory()) {
+      return Collections.emptyList();
+    }
+    final File[] files = (filter == null) ? directory.listFiles() : directory.listFiles(filter);
+    return (files != null) ? Arrays.asList(files) : Collections.emptyList();
+  }
+
+  private static File prepareDirectory(File directory) {
+    if (directory == null) {
+      return null;
+    }
+
+    if (directory.exists() || directory.mkdirs()) {
+      return directory;
+    }
+
+    // TODO: Couldn't create directory. Log? Throw?
+    return null;
+  }
+
+  private static void writeTextFile(File file, String text) {
+    try (OutputStreamWriter writer = new OutputStreamWriter(new FileOutputStream(file), UTF_8)) {
+      writer.write(text);
+    } catch (IOException e) {
+      // TODO: Exception writing file to disk. Log? Throw?
+    }
+  }
+
+  private static String readTextFile(File file) {
+    final byte[] readBuffer = new byte[8192];
+    final ByteArrayOutputStream bos = new ByteArrayOutputStream();
+    try (FileInputStream fileInput = new FileInputStream(file)) {
+      int read;
+      while ((read = fileInput.read(readBuffer)) > 0) {
+        bos.write(readBuffer, 0, read);
+      }
+      return new String(bos.toByteArray(), UTF_8);
+    } catch (IOException e) {
+      return null;
+    }
+  }
+
+  private static void recursiveDelete(File file) {
```

Comment:
These methods already deleted elsewhere in the SDK, correct? So is the plan to delete the other versions?

Reply:
Ultimately, yes. My goal in the long term is to reduce "Utils" classes, so for now I went with the Go proverb "A little copying is better than a little dependency."